pfcpiface/bess.go

```diff
@@ -985,6 +1005,109 @@ func (b *bess) delCounter(ctx context.Context, done chan<- bool, ctrID uint32, c
 	}()
 }
 
+func (b *bess) processSliceMeter(ctx context.Context, any *anypb.Any, method upfMsgType) {
+	if method != upfMsgTypeAdd && method != upfMsgTypeDel && method != upfMsgTypeClear {
+		log.Errorln("Invalid method name: ", method)
+		return
+	}
+
+	methods := [...]string{"add", "add", "delete", "clear"}
+
+	_, err := b.client.ModuleCommand(
+		ctx, &pb.CommandRequest{
+			Name: "sliceMeter",
+			Cmd:  methods[method],
+			Arg:  any,
+		},
+	)
+	if err != nil {
+		log.Errorln("sliceMeter method failed!:", err)
+	}
+}
+
+func (b *bess) addSliceMeter(ctx context.Context, done chan<- bool, meterConfig SliceMeterConfig) {
+	go func() {
+		var (
+			any                           *anypb.Any
+			err                           error
+			cir, pir, cbs, ebs, pbs, gate uint64
+		)
+
+		// Uplink slice meter config
+		if meterConfig.UplinkRateBps != 0 {
+			gate = sliceMeterGateMeter
+			cir = 1                             // Mark all traffic as yellow
+			pir = meterConfig.UplinkRateBps / 8 // bit/s to byte/s
+		} else {
+			gate = sliceMeterGateUnmeter
+		}
+		if meterConfig.UplinkBurstBytes != 0 {
+			cbs = 1 // Mark all traffic as yellow
+			pbs = meterConfig.UplinkBurstBytes
+			ebs = 0 // Unused
+		} else {
+			cbs = DefaultBurstSize
+			pbs = DefaultBurstSize
+			ebs = 0 // Unused
+		}
+		q := &pb.QosCommandAddArg{
+			Gate:                    gate,
+			Cir:                     cir, /* committed info rate */
+			Pir:                     pir, /* peak info rate */
+			Cbs:                     cbs, /* committed burst size */
+			Pbs:                     pbs, /* Peak burst size */
+			Ebs:                     ebs, /* Excess burst size */
+			AdjustMeterPacketLength: -14, /* Exclude Ethernet header */
+			Fields: []*pb.FieldData{
+				intEnc(uint64(access)), /* Source interface */
+			},
+		}
+		any, err = anypb.New(q)
+		if err != nil {
+			log.Errorln("Error marshalling the rule", q, err)
+			return
+		}
+		b.processSliceMeter(ctx, any, upfMsgTypeAdd)
+
+		// Downlink slice meter config
+		if meterConfig.DownlinkRateBps != 0 {
+			gate = sliceMeterGateMeter
+			cir = 1                               // Mark all traffic as yellow
+			pir = meterConfig.DownlinkRateBps / 8 // bit/s to byte/s
+		} else {
+			gate = sliceMeterGateUnmeter
+		}
+		if meterConfig.DownlinkBurstBytes != 0 {
+			cbs = 1 // Mark all traffic as yellow
+			pbs = meterConfig.DownlinkBurstBytes
+			ebs = 0 // Unused
+		} else {
+			cbs = DefaultBurstSize
+			pbs = DefaultBurstSize
+			ebs = 0 // Unused
+		}
+		q = &pb.QosCommandAddArg{
+			Gate:                    gate,
+			Cir:                     cir, /* committed info rate */
+			Pir:                     pir, /* peak info rate */
+			Cbs:                     cbs, /* committed burst size */
+			Pbs:                     pbs, /* Peak burst size */
+			Ebs:                     ebs, /* Excess burst size */
+			AdjustMeterPacketLength: -50, /* Exclude Ethernet,IP,UDP,GTP header */
```

Comment:
The assumption here is that all downlink traffic is GTP encaped. Is this correct in all cases?

Reply:
Access - N3 - encap
Core - N9 - encap
Core - N6 - not encap

Reply:
mhh, that means we're slightly off for traffic going to the core, because we can only subtract one value (either -14 or -50).

Reply:
For what I know, we don't have to worry about N9 for now, as we don't have plans to support I-UPF in Aether for the foreseeable future. @thakurajayL @badhrinathpa please confirm.